### ISSUE TITLE
Per-file strict optional doesn't play well with subtype cache

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#5603]
Consider this test case:
```
[case testPerFileStrictOptionalTricky]
# flags: --config-file tmp/mypy.ini
import a
[file a.py]
from typing import List, Optional
import b
x: List[int]
y: List[Optional[int]]
x = y  # E: Incompatible types in assignment (expression has type "List[Optional[int]]", variable has type "List[int]")
[file b.py]
from typing import List, Optional
x: List[int]
y: List[Optional[int]]
x = y

[file mypy.ini]
[[mypy]
strict_optional = True
[[mypy-b]
strict_optional = False
[builtins fixtures/list.pyi]
[out]
```
The error in `a.py` is not detected because the _same_ two types are first compared in `b.py` and the  positive result is stored in the subtype cache.

After studying a bit the code around per-file strict optional, it looks quite problematic. There is `TypeChecker.should_suppress_optional_error()` with some questionable code, but it actually doesn't work because `self.suppress_none_errors` is `False` even if strict optional is disabled. In fact, the strict optional errors are suppressed because `SubtypeVisitor` has code that depends on `experiments.STRICT_OPTIONAL`.

In real world code, the false negatives caused by this are rare, but I think this is worth fixing because the code is fragile.

**Comment by JukkaL** [comment]
Would it help if we added the value of `experiments.STRICT_OPTIONAL` (the name of which is a misnomer) in the subtype cache key?

**Comment by ilevkivskyi** [comment]
Yes, but I think we already have many flags there, so I would avoid adding them (plus it may affect performance).

**Comment by JukkaL** [comment]
Have you a better way to fix the issue?

I don't see adding a new flag to a cache key as a problem. It will be used only in a few places, and we just need to document what it means and why we include it.

The performance impact is easy enough to measure/estimate once we have an implementation.
